### PR TITLE
SVC-5148 Add method to cycle alerts on/off on schedule

### PR DESCRIPTION
**Pull request by bsper2**

### CODE REVIEW

**Comment by bsper2** [comment]
This has been tested on metrics-test.ncsa.illinois.edu, metrics02.ncsa.illinois.edu, and metrics01.ncsa.illinois.edu for the past week, and alerts pause as intended.

**Comment by billglick** [comment]
Just thinking out loud...

At some point it might make sense for the pause information to be driven somehow as an exported resource via PuppetDB by the host that is configured to do a scheduled maintenance task. 

As an example `profile_update_os::kernel_upgrade` (from https://github.com/ncsa/puppet-profile_update_os ) has the option to set a kernel upgrade and reboot via CRON. Perhaps that class should be able to export the alert on/off code for each host that runs this.

File: manifests/alert_cycle.pp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#
+# @param cycle_alerts
+#   Enable or disable the cycling of alerts
+#
+# @param script_path
+#   Path to the alert_toggle.sh script
+#   
+class profile_metrics_alerting::alert_cycle (
+  Hash $crons,
+  Boolean $cycle_alerts,
```
Comment (billglick) [inline manifests/alert_cycle.pp]: I'd prefer to name this variable something like `enable_cycle_alerts`.

It was not obvious to me reading about this parameter that this was a boolean that toggles whether or not the cycling of alerts was enabled.

File: README.md [hunk 2]
```diff
@@ -49,6 +49,29 @@ profile_metrics_alerting::tools::whitelabel_subtitle: "Managed by ORGANIZATION G
...
+    minute: 55
+    weekday: 0
+  "Start mforgehn1 alerts":
+    command: "start tag \"mforgehn1\""
+    hour: 0
+    minute: 45
+    weekday: 1
+profile_metrics_alerting::alert_cycle::cycle_alerts: true
+```
+You'll want to define your crons for `profile_metrics_alerting::alert_cycle::crons` in any role that uses this repo (For example if you have a primary/secondary role). This is because if you swap roles you want the alert pausing/starting crons to be removed from the secondary role so you don't get alerts enabled in multiple places leading to duplicate alerts.
```
Comment (billglick) [inline README.md:68]: This isn't obvious to me how this works for secondary servers. It seems like the secondary roles should never `start` any alerts.
Perhaps I'm misinterpreting what this line is saying.

File: README.md [hunk 3]
```diff
@@ -49,6 +49,29 @@ profile_metrics_alerting::tools::whitelabel_subtitle: "Managed by ORGANIZATION G
...
+    weekday: 1
+profile_metrics_alerting::alert_cycle::cycle_alerts: true
+```
+You'll want to define your crons for `profile_metrics_alerting::alert_cycle::crons` in any role that uses this repo (For example if you have a primary/secondary role). This is because if you swap roles you want the alert pausing/starting crons to be removed from the secondary role so you don't get alerts enabled in multiple places leading to duplicate alerts. 
+
+### class profile_metrics_alerting::ssh (
+-  Array[String] $metrics_node_ips,
+-  String $sshkey_pub,
+-  String $sshkey_priv,
+-  String $sshkey_type,
```
Comment (billglick) [inline README.md:90]: Should all of these parameters (including the others not tagged by this comment) be moved lower into the README file under the `Reference` heading?

File: files/root/grafana_tools/bulk_alert/alert_toggle.sh [hunk 4]
```diff
@@ -4,6 +4,7 @@ action=$1
 grouping=$2
 tag_or_id=$3
 
+pushd $(dirname "$0")
```
Comment (billglick) [inline files/root/grafana_tools/bulk_alert/alert_toggle.sh:7]: Can you explain why this was added?

I'm sure it is probably all good, but just curious why the change.

File: manifests/alert_cycle.pp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  String $script_path,
+) {
+
+  Cron {
+    user => 'root',
+  }
+
+  if ($cycle_alerts) {
+    $crons.each | $k, $v | {
```
Comment (billglick) [inline manifests/alert_cycle.pp:24]: This is nice code. I want to steal this for other profiles.

**Review by billglick — CHANGES_REQUESTED** [review]
This looks really good. 
I did have a few minor comments, suggestions, questions. But generally this looks good.

File: files/root/grafana_tools/bulk_alert/alert_toggle.sh (hunk 4 above)
Comment (billglick) [reply files/root/grafana_tools/bulk_alert/alert_toggle.sh:7]: Also... make sure this gets noted in the parent GitHub repo as an issue/change:
- https://github.com/jdmaloney/grafana_tools

File: README.md (hunk 2 above)
Comment (bsper2) [reply README.md:68]: I was trying to make the alert cycling properly handle someone swapping between a node between the metrics_primary and metrics_secondary roles.

If a node goes from primary to secondary, we want the crons that cycle the alerts to go away (or at least disable themselves). Since crons that aren't managed by puppet do not get purged, we need a way to tell Puppet the name of each cron we want to purge. So my comment here is talking about how you want to define all your crons in both the metrics_primary and metrics_secondary role, like how I did here : https://git.ncsa.illinois.edu/ici/asd/asd-pup-control/-/compare/main...SVC-5148%2Fcycle_alerts

It's ugly and I don't really like doing it that way, but I didn't see any immediate better ways to handle it. Tested seeing if I could add a standard prefix to all the crons, and tell puppet to delete all crons that start with 'alert_cycle_*', but it doesn't look like I can give wildcards in the resource title like that since I got errors.

I'll add some clarification to the README to better explain it.

File: files/root/grafana_tools/bulk_alert/alert_toggle.sh (hunk 4 above)
Comment (bsper2) [reply files/root/grafana_tools/bulk_alert/alert_toggle.sh:7]: Same reason for https://github.com/jdmaloney/grafana_tools/issues/2, in that the script was breaking since it was running in a different directory when ran via cron.

The alert_toggle.sh script had more than just a single `source ./config`, since it was also loading the json files from `.`, so I opted to have the script pushd and popd back instead of editing the path for each of those lines.

I did see that version of grafana_tools, and we also have another copy over in https://git.ncsa.illinois.edu/ici-monitoring/grafana-tools. So wasn't sure which is our authoritative copy. They have separate commit histories so syncing them might be a bit more ugly than usual. Since one is in JD's personal github space maybe best to use our gitlab version as authoritative.

File: manifests/alert_cycle.pp (hunk 5 above)
Comment (bsper2) [reply manifests/alert_cycle.pp:24]: It's mostly stolen from here: https://github.com/ncsa/puppet-profile_audit/blob/main/manifests/root_equivalence.pp

I just added some lines to have the crons get deleted when not needed, and some lines to customize the command given so I don't need to give it the script path each time.

File: README.md (hunk 2 above)
Comment (billglick) [reply README.md:68]: The current status of alerting on the `metrics_secondary` servers is kind of messy and not obvious how they work.

Currently if the `profile_metrics_alerting::tools::backup_and_sync_enable` is `TRUE` (which it is on `metrics_primary`) then that script does a `./alert_toggle.sh pause tag all` as a last step on the server that it syncs to (the destination). Currently this has the result of pausing all alerts on `metrics02`. But now the changes in this branch are now unpausing alerts for some specific hosts on metrics02 (at least until the next sync gets pushed to it).

I do not think we ever want any alerts enabled on `metrics02` unless `metrics01` fails and we need to temporarily make `metrics02` take the `metrics` CNAME and switch to the primary role.

But I can see that perhaps we do want to enable and test alerts from `metrics-test`, which unfortunately also uses the `metrics_secondary` role.

Perhaps we really need to switch up these roles more so that they are more clear what they are doing. Until now, the `metrics_primary` role primarily had the following distinctions about it:
- use `metrics.ncsa.illinois.edu` for its web configuration
- enable `profile_metrics_alerting::tools::backup_and_sync_enable` to sync data to `metrics02`

File: files/root/grafana_tools/bulk_alert/alert_toggle.sh (hunk 4 above)
Comment (billglick) [reply files/root/grafana_tools/bulk_alert/alert_toggle.sh:7]: OK, that makes sense.  If we're doing that, then let's also:
- update `~/grafana_tools/backup_and_sync/sync_grafana.sh` to no longer require the `cd` in its `alert_toggle.sh` lines
- apply the same fix to `grafana_tools/ldap_sync/sync.sh` and its CRONs from puppet
- apply the same fix to `grafana_tools/backup_and_sync/sync_grafana` and its CRONs from puppet

I thought JD told me that https://github.com/jdmaloney/grafana_tools was what should be used, but now I'm questioning which is most recent. 😞

File: manifests/alert_cycle.pp (hunk 1 above)
Comment (bsper2) [reply manifests/alert_cycle.pp]: Good point, I've renamed it to `enable_cycle_alerts`

File: README.md (hunk 2 above)
Comment (bsper2) [reply README.md:68]: > But now the changes in this branch are now unpausing alerts for some specific hosts on metrics02 (at least until the next sync gets pushed to it)

This part isn't a problem, the crons for pausing/un-pausing never set foot on metrics-test or metrics02. By default `profile_metrics_alerting::alert_cycle::enable_cycle_alerts` is set to false, and the only place I set that to true is in `data/role/metrics_primary.yaml`. The annoying and confusing part is that I do still define the crons themselves over in `data/role/metrics_secondary.yaml`, but that's only so that I can give Puppet a list of crons that it should delete if they happen to exist. Which they would exist if a node ever changed from role `metrics_primary` to `metrics_secondary`.

If we ever wanted to do some testing where metrics-test would have the crons to cycle alerts on, we could keep it in the same `metrics_secondary` role and then set `profile_metrics_alerting::alert_cycle::enable_cycle_alerts: true` in `data/nodes/metrics-test.ncsa.illinois.edu.yaml`

**Comment by bsper2** [comment]
Scripts have been updated to include pushd/popd, and crons set in manifests/tools.pp have been updated to remove the no longer necessary cd in the command.

Tested on metrics-test, metrics02, and metrics01

**Review by billglick — APPROVED** [review]
Looks good
